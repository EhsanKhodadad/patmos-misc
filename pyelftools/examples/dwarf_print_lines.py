#----------------------------------------------------------------------
# This script reads DWARF debug info from an ELF file and
# prints out function names and line number infos for every
# address that is given as an input. 
#
# This script is designed to be used with objdump or similar tools.
# To annotate the objdump output with line infos, use 
#
# objdump -d <binary> | dwarf_print_lines.py <binary>
# 
# Author: Stefan Hepp <hepp@complang.tuwien.ac.at>
# This code is in the public domain
#----------------------------------------------------------------------

from __future__ import print_function
import sys
import os
import argparse
import re

# We do not care about closed stdout pipe, just abort 
import signal
signal.signal(signal.SIGPIPE, signal.SIG_DFL)
signal.signal(signal.SIGINT,  signal.SIG_DFL)

# If elftools is not installed, maybe we're running from the root or examples
# dir of the source distribution
try:
    import elftools
except ImportError:
    sys.path.extend(['.', '..'])

from elftools.common.py3compat import itervalues, maxint, bytes2str
from elftools.elf.elffile import ELFFile


def decode_funcname(funcinfo, address):
    for info in funcinfo:
        lowpc = info[0]
        highpc = info[1]
        if lowpc <= address <= highpc:
            return info[2]
    return None


def decode_file_line(lineinfo, address):
    for info in lineinfo:
        stateaddress = info[0]
        nextaddress = info[1]
        if stateaddress <= address < nextaddress:
            filename = info[2]
            line = info[3]
            return filename, line
    return None, None

def load_lineinfo(dwarfinfo):
    # Load all line infos and all function names into lists
    # Lines contains tuples of startaddr, endaddress, filename, line
    lines = []
    # Functions contains tuples of lowpc, highpc, function
    functions = []
    for CU in dwarfinfo.iter_CUs():
        # First, look at line programs to find the file/line map
        lineprog = dwarfinfo.line_program_for_CU(CU)
        prevstate = None
        for entry in lineprog.get_entries():
            # We're interested in those entries where a new state is assigned
            state = entry.state
            if state is None: continue
            if prevstate and prevstate.address <= state.address and not prevstate.end_sequence:
                file_entry = lineprog['file_entry'][prevstate.file - 1]
                if file_entry.dir_index == 0:
                    # current directory
                    # TODO get directory of source file and prepend it 
                    filename = './%s' % (bytes2str(file_entry.name))
                else:
                    filename = '%s/%s' % (
                        bytes2str(lineprog['include_directory'][file_entry.dir_index - 1]),
                        bytes2str(file_entry.name))
                line = prevstate.line
                info = prevstate.address, state.address, filename, line
                lines.append( info )
            prevstate = state
        # Go over all DIEs in the DWARF information. Note that
        # this simplifies things by disregarding subprograms that may have 
        # split address ranges.
        for DIE in CU.iter_DIEs():
            try:
                if DIE.tag == 'DW_TAG_subprogram':
                    lowpc = DIE.attributes['DW_AT_low_pc'].value
                    highpc = DIE.attributes['DW_AT_high_pc'].value
                    function = DIE.attributes['DW_AT_name'].value
                    info = lowpc, highpc, bytes2str(function)
                    functions.append( info )
            except KeyError:
                continue
    return lines, functions

def print_lineinfos(lineinfo, functioninfo, input, printAllLines, printInput = True):
    # Try to match first column as hex PC
    ppc = re.compile("([a-fA-F0-9]+)[ \t\n]")
    # Try to match objdump format
    opc = re.compile(" +([a-f0-9]+):")
    
    lastLineNr = None
    lastFilename = None
    lastFunction = None

    for line in input:
        
        # Check if we have a PC at the beginning of the line
        PC = None
        for r in [ppc, opc]: 
            m = r.match(line)
            if m:
                PC = int(m.group(1), 16)
                break
        if PC is None:
            if printInput: print(line, end="")
            continue
        
        function = decode_funcname(functioninfo, PC)
        filename, linenr = decode_file_line(lineinfo, PC)
        
        if printAllLines:
            print('%s:%s, %s():' % (filename, linenr, function))
        else:
            if function and lastFunction != function:
                print("%s():" % function)
            if (filename or linenr) and (lastLineNr != linenr or lastFilename != filename):
        	   print("%s:%s" % (filename, linenr))
                
        lastFunction = function
        lastFilename = filename
        lastLineNr = linenr

        if printInput: print(line, end="")

def load_dwarfinfo(filename):
    elffile = ELFFile(filename)
    
    if not elffile.has_dwarf_info():
        print(filename.name + ': ELF file has no DWARF info!')
        sys.exit(1)
    
    # get_dwarf_info returns a DWARFInfo context object, which is the
    # starting point for all DWARF-based processing in pyelftools.
    dwarfinfo = elffile.get_dwarf_info()
    
    return dwarfinfo


parser = argparse.ArgumentParser(description='Display ELF debug infos.')
parser.add_argument('elffile', type=file, help='ELF file containing debug infos')
parser.add_argument('-a', '--all', action='store_true', help='Print line numbers for all found PCs, not only when it changes')
parser.add_argument('-l', '--lines', help='Display line numbers for the given input file (defaults to stdin)')

args = parser.parse_args()

# Load the ELF file
dwarfinfo = load_dwarfinfo(args.elffile)
# Load all dwarf line number and function infos into a lookup table
lineinfo, funcinfo = load_lineinfo(dwarfinfo)

if args.lines and args.lines != "-": 
    input = None
    try:
	input = open(args.lines, 'r')
    except:
	print("Could not open input file " + args.lines + ": " + e)
	sys.exit(1)
    
    print_lineinfos(lineinfo, funcinfo, input, args.all)
    
    input.close()
    
else:
    print_lineinfos(lineinfo, funcinfo, sys.stdin, args.all)

