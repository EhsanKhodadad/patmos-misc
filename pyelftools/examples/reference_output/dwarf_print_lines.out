
sample_exe64.elf:     file format elf64-x86-64


Disassembly of section .init:

00000000004003c0 <_init>:
  4003c0:	48 83 ec 08          	sub    $0x8,%rsp
  4003c4:	e8 63 00 00 00       	callq  40042c <call_gmon_start>
  4003c9:	e8 f2 00 00 00       	callq  4004c0 <frame_dummy>
  4003ce:	e8 ed 01 00 00       	callq  4005c0 <__do_global_ctors_aux>
  4003d3:	48 83 c4 08          	add    $0x8,%rsp
  4003d7:	c3                   	retq   

Disassembly of section .plt:

00000000004003d8 <__libc_start_main@plt-0x10>:
  4003d8:	ff 35 12 0c 20 00    	pushq  0x200c12(%rip)        # 600ff0 <_GLOBAL_OFFSET_TABLE_+0x8>
  4003de:	ff 25 14 0c 20 00    	jmpq   *0x200c14(%rip)        # 600ff8 <_GLOBAL_OFFSET_TABLE_+0x10>
  4003e4:	0f 1f 40 00          	nopl   0x0(%rax)

00000000004003e8 <__libc_start_main@plt>:
  4003e8:	ff 25 12 0c 20 00    	jmpq   *0x200c12(%rip)        # 601000 <_GLOBAL_OFFSET_TABLE_+0x18>
  4003ee:	68 00 00 00 00       	pushq  $0x0
  4003f3:	e9 e0 ff ff ff       	jmpq   4003d8 <_init+0x18>

Disassembly of section .text:

../sysdeps/x86_64/elf/start.S:65
0000000000400400 <_start>:
  400400:	31 ed                	xor    %ebp,%ebp
../sysdeps/x86_64/elf/start.S:81
  400402:	49 89 d1             	mov    %rdx,%r9
../sysdeps/x86_64/elf/start.S:83
  400405:	5e                   	pop    %rsi
../sysdeps/x86_64/elf/start.S:84
  400406:	48 89 e2             	mov    %rsp,%rdx
../sysdeps/x86_64/elf/start.S:86
  400409:	48 83 e4 f0          	and    $0xfffffffffffffff0,%rsp
../sysdeps/x86_64/elf/start.S:88
  40040d:	50                   	push   %rax
../sysdeps/x86_64/elf/start.S:92
  40040e:	54                   	push   %rsp
../sysdeps/x86_64/elf/start.S:106
  40040f:	49 c7 c0 20 05 40 00 	mov    $0x400520,%r8
../sysdeps/x86_64/elf/start.S:107
  400416:	48 c7 c1 30 05 40 00 	mov    $0x400530,%rcx
../sysdeps/x86_64/elf/start.S:109
  40041d:	48 c7 c7 ec 04 40 00 	mov    $0x4004ec,%rdi
../sysdeps/x86_64/elf/start.S:113
  400424:	e8 bf ff ff ff       	callq  4003e8 <__libc_start_main@plt>
../sysdeps/x86_64/elf/start.S:116
  400429:	f4                   	hlt    
  40042a:	90                   	nop
  40042b:	90                   	nop

000000000040042c <call_gmon_start>:
  40042c:	48 83 ec 08          	sub    $0x8,%rsp
  400430:	48 8b 05 a9 0b 20 00 	mov    0x200ba9(%rip),%rax        # 600fe0 <_DYNAMIC+0x1a0>
  400437:	48 85 c0             	test   %rax,%rax
  40043a:	74 02                	je     40043e <call_gmon_start+0x12>
  40043c:	ff d0                	callq  *%rax
  40043e:	48 83 c4 08          	add    $0x8,%rsp
  400442:	c3                   	retq   
  400443:	90                   	nop
  400444:	90                   	nop
  400445:	90                   	nop
  400446:	90                   	nop
  400447:	90                   	nop
  400448:	90                   	nop
  400449:	90                   	nop
  40044a:	90                   	nop
  40044b:	90                   	nop
  40044c:	90                   	nop
  40044d:	90                   	nop
  40044e:	90                   	nop
  40044f:	90                   	nop

0000000000400450 <__do_global_dtors_aux>:
  400450:	55                   	push   %rbp
  400451:	48 89 e5             	mov    %rsp,%rbp
  400454:	53                   	push   %rbx
  400455:	48 83 ec 08          	sub    $0x8,%rsp
  400459:	80 3d c0 0b 20 00 00 	cmpb   $0x0,0x200bc0(%rip)        # 601020 <completed.6159>
  400460:	75 4b                	jne    4004ad <__do_global_dtors_aux+0x5d>
  400462:	b8 30 0e 60 00       	mov    $0x600e30,%eax
  400467:	48 8b 15 ba 0b 20 00 	mov    0x200bba(%rip),%rdx        # 601028 <dtor_idx.6161>
  40046e:	48 2d 28 0e 60 00    	sub    $0x600e28,%rax
  400474:	48 c1 f8 03          	sar    $0x3,%rax
  400478:	48 8d 58 ff          	lea    -0x1(%rax),%rbx
  40047c:	48 39 da             	cmp    %rbx,%rdx
  40047f:	73 25                	jae    4004a6 <__do_global_dtors_aux+0x56>
  400481:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)
  400488:	48 8d 42 01          	lea    0x1(%rdx),%rax
  40048c:	48 89 05 95 0b 20 00 	mov    %rax,0x200b95(%rip)        # 601028 <dtor_idx.6161>
  400493:	ff 14 c5 28 0e 60 00 	callq  *0x600e28(,%rax,8)
  40049a:	48 8b 15 87 0b 20 00 	mov    0x200b87(%rip),%rdx        # 601028 <dtor_idx.6161>
  4004a1:	48 39 da             	cmp    %rbx,%rdx
  4004a4:	72 e2                	jb     400488 <__do_global_dtors_aux+0x38>
  4004a6:	c6 05 73 0b 20 00 01 	movb   $0x1,0x200b73(%rip)        # 601020 <completed.6159>
  4004ad:	48 83 c4 08          	add    $0x8,%rsp
  4004b1:	5b                   	pop    %rbx
  4004b2:	c9                   	leaveq 
  4004b3:	c3                   	retq   
  4004b4:	66 66 66 2e 0f 1f 84 	data32 data32 nopw %cs:0x0(%rax,%rax,1)
  4004bb:	00 00 00 00 00 

00000000004004c0 <frame_dummy>:
  4004c0:	55                   	push   %rbp
  4004c1:	48 83 3d 6f 09 20 00 	cmpq   $0x0,0x20096f(%rip)        # 600e38 <__JCR_END__>
  4004c8:	00 
  4004c9:	48 89 e5             	mov    %rsp,%rbp
  4004cc:	74 1a                	je     4004e8 <frame_dummy+0x28>
  4004ce:	b8 00 00 00 00       	mov    $0x0,%eax
  4004d3:	48 85 c0             	test   %rax,%rax
  4004d6:	74 10                	je     4004e8 <frame_dummy+0x28>
  4004d8:	bf 38 0e 60 00       	mov    $0x600e38,%edi
  4004dd:	49 89 c3             	mov    %rax,%r11
  4004e0:	c9                   	leaveq 
  4004e1:	41 ff e3             	jmpq   *%r11
  4004e4:	0f 1f 40 00          	nopl   0x0(%rax)
  4004e8:	c9                   	leaveq 
  4004e9:	c3                   	retq   
  4004ea:	90                   	nop
  4004eb:	90                   	nop

main():
./z.c:2
00000000004004ec <main>:
  4004ec:	55                   	push   %rbp
  4004ed:	48 89 e5             	mov    %rsp,%rbp
  4004f0:	89 7d fc             	mov    %edi,-0x4(%rbp)
  4004f3:	48 89 75 f0          	mov    %rsi,-0x10(%rbp)
./z.c:3
  4004f7:	48 8b 45 f0          	mov    -0x10(%rbp),%rax
  4004fb:	89 c2                	mov    %eax,%edx
  4004fd:	8b 05 15 0b 20 00    	mov    0x200b15(%rip),%eax        # 601018 <glob>
  400503:	8d 04 02             	lea    (%rdx,%rax,1),%eax
  400506:	89 05 0c 0b 20 00    	mov    %eax,0x200b0c(%rip)        # 601018 <glob>
./z.c:4
  40050c:	8b 05 06 0b 20 00    	mov    0x200b06(%rip),%eax        # 601018 <glob>
  400512:	03 45 fc             	add    -0x4(%rbp),%eax
./z.c:5
  400515:	c9                   	leaveq 
  400516:	c3                   	retq   
  400517:	90                   	nop
  400518:	90                   	nop
  400519:	90                   	nop
  40051a:	90                   	nop
  40051b:	90                   	nop
  40051c:	90                   	nop
  40051d:	90                   	nop
  40051e:	90                   	nop
  40051f:	90                   	nop

__libc_csu_fini():
./elf-init.c:140
0000000000400520 <__libc_csu_fini>:
  400520:	f3 c3                	repz retq 
  400522:	66 66 66 66 66 2e 0f 	data32 data32 data32 data32 nopw %cs:0x0(%rax,%rax,1)
  400529:	1f 84 00 00 00 00 00 

__libc_csu_init():
./elf-init.c:106
0000000000400530 <__libc_csu_init>:
  400530:	48 89 6c 24 d8       	mov    %rbp,-0x28(%rsp)
  400535:	4c 89 7c 24 f8       	mov    %r15,-0x8(%rsp)
./elf-init.c:123
  40053a:	48 8d 2d d3 08 20 00 	lea    0x2008d3(%rip),%rbp        # 600e14 <__init_array_end>
  400541:	4c 8d 3d cc 08 20 00 	lea    0x2008cc(%rip),%r15        # 600e14 <__init_array_end>
./elf-init.c:106
  400548:	4c 89 64 24 e0       	mov    %r12,-0x20(%rsp)
  40054d:	4c 89 6c 24 e8       	mov    %r13,-0x18(%rsp)
  400552:	4c 89 74 24 f0       	mov    %r14,-0x10(%rsp)
  400557:	48 89 5c 24 d0       	mov    %rbx,-0x30(%rsp)
  40055c:	48 83 ec 38          	sub    $0x38,%rsp
./elf-init.c:123
  400560:	4c 29 fd             	sub    %r15,%rbp
./elf-init.c:106
  400563:	41 89 fe             	mov    %edi,%r14d
  400566:	49 89 f5             	mov    %rsi,%r13
./elf-init.c:123
  400569:	48 c1 fd 03          	sar    $0x3,%rbp
./elf-init.c:106
  40056d:	49 89 d4             	mov    %rdx,%r12
./elf-init.c:120
  400570:	e8 4b fe ff ff       	callq  4003c0 <_init>
./elf-init.c:123
  400575:	48 85 ed             	test   %rbp,%rbp
  400578:	74 1c                	je     400596 <__libc_csu_init+0x66>
  40057a:	31 db                	xor    %ebx,%ebx
  40057c:	0f 1f 40 00          	nopl   0x0(%rax)
./elf-init.c:124
  400580:	4c 89 e2             	mov    %r12,%rdx
  400583:	4c 89 ee             	mov    %r13,%rsi
  400586:	44 89 f7             	mov    %r14d,%edi
  400589:	41 ff 14 df          	callq  *(%r15,%rbx,8)
./elf-init.c:123
  40058d:	48 83 c3 01          	add    $0x1,%rbx
  400591:	48 39 eb             	cmp    %rbp,%rbx
  400594:	72 ea                	jb     400580 <__libc_csu_init+0x50>
./elf-init.c:125
  400596:	48 8b 5c 24 08       	mov    0x8(%rsp),%rbx
  40059b:	48 8b 6c 24 10       	mov    0x10(%rsp),%rbp
  4005a0:	4c 8b 64 24 18       	mov    0x18(%rsp),%r12
  4005a5:	4c 8b 6c 24 20       	mov    0x20(%rsp),%r13
  4005aa:	4c 8b 74 24 28       	mov    0x28(%rsp),%r14
  4005af:	4c 8b 7c 24 30       	mov    0x30(%rsp),%r15
  4005b4:	48 83 c4 38          	add    $0x38,%rsp
  4005b8:	c3                   	retq   
  4005b9:	90                   	nop
  4005ba:	90                   	nop
  4005bb:	90                   	nop
  4005bc:	90                   	nop
  4005bd:	90                   	nop
  4005be:	90                   	nop
  4005bf:	90                   	nop

00000000004005c0 <__do_global_ctors_aux>:
  4005c0:	55                   	push   %rbp
  4005c1:	48 89 e5             	mov    %rsp,%rbp
  4005c4:	53                   	push   %rbx
  4005c5:	48 83 ec 08          	sub    $0x8,%rsp
  4005c9:	48 8b 05 48 08 20 00 	mov    0x200848(%rip),%rax        # 600e18 <__CTOR_LIST__>
  4005d0:	48 83 f8 ff          	cmp    $0xffffffffffffffff,%rax
  4005d4:	74 19                	je     4005ef <__do_global_ctors_aux+0x2f>
  4005d6:	bb 18 0e 60 00       	mov    $0x600e18,%ebx
  4005db:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)
  4005e0:	48 83 eb 08          	sub    $0x8,%rbx
  4005e4:	ff d0                	callq  *%rax
  4005e6:	48 8b 03             	mov    (%rbx),%rax
  4005e9:	48 83 f8 ff          	cmp    $0xffffffffffffffff,%rax
  4005ed:	75 f1                	jne    4005e0 <__do_global_ctors_aux+0x20>
  4005ef:	48 83 c4 08          	add    $0x8,%rsp
  4005f3:	5b                   	pop    %rbx
  4005f4:	c9                   	leaveq 
  4005f5:	c3                   	retq   
  4005f6:	90                   	nop
  4005f7:	90                   	nop

Disassembly of section .fini:

00000000004005f8 <_fini>:
  4005f8:	48 83 ec 08          	sub    $0x8,%rsp
  4005fc:	e8 4f fe ff ff       	callq  400450 <__do_global_dtors_aux>
  400601:	48 83 c4 08          	add    $0x8,%rsp
  400605:	c3                   	retq   
